File: EPI/15.1/swift/15.1.swift
public class Node {
    let value: Int
    let left: Node?
    let right: Node?

    public init(value: Int, left: Node? = nil, right: Node? = nil) {
        self.value = value
        self.left = left
        self.right = right
    }
}

extension Node {
    public func isBST() -> Bool {
        var result = true
        var prev = Int.min

        depthFirstSearch { node in
            result = result && node.value >= prev
            prev = node.value
        }

        return result
    }

    private func depthFirstSearch(_ process: (Node) -> ()) {
        left?.depthFirstSearch(process)
        process(self)
        right?.depthFirstSearch(process)
    }
}

File: EPI/Binary Search Trees/Test if a binary tree satisfies the BST property/typescript/15.1.ts
export interface Node {
    value: number;
    left?: Node;
    right?: Node;
}

export function isBST(root: Node): boolean {
    return isBSTRec(root, Number.NEGATIVE_INFINITY, Number.POSITIVE_INFINITY);
}

function isBSTRec(root: Node | undefined, min: number, max: number): boolean {
    if (!root) {
        return true;
    }

    const {value, left, right} = root;

    if (!(min <= value && value <= max)) {
        return false;
    }

    return isBSTRec(left, min, Math.min(max, value)) &&
        isBSTRec(right, Math.max(min, value), max);
}
